File: frontend/trader/src/lib/api/insurance.ts
/**
 * Trade Insurance API client.
 * Backed by /api/v1/insurance/* — see backend/services/gateway/src/api/insurance.py
 */
import api from './client';

export type InsuranceTier = 'basic' | 'advanced' | 'pro' | 'elite';

export interface TierQuote {
  tier: InsuranceTier;
  fee: number;
  coverage_pct: number;
  max_cap: number;
  estimated_refund: number;
  risk_score: number;
}

export interface QuoteRequest {
  account_id: string;
  symbol: string;
  side: 'buy' | 'sell';
  lots: number;
  leverage?: number;
  stop_loss?: number;
  take_profit?: number;
}

export interface ActivateResponse {
  policy_id: string;
  fee_charged: string;
  status: 'active';
}

export interface PolicyOut {
  id: string;
  position_id: string | null;
  instrument_symbol: string | null;
  tier: InsuranceTier;
  fee: string;
  coverage_pct: string;
  max_cap: string;
  status: 'active' | 'claimed' | 'expired' | 'denied';
  activated_at: string;
  settled_at: string | null;
}

export interface ClaimOut {
  id: string;
  policy_id: string;
  loss_amount: string;
  claim_amount: string;
  paid_at: string;
}

export const insuranceApi = {
  quote: (body: QuoteRequest) => api.post<TierQuote[]>('/insurance/quote', body),
  activate: (position_id: string, tier: InsuranceTier) =>
    api.post<ActivateResponse>('/insurance/activate', { position_id, tier }),
  active: () => api.get<PolicyOut[]>('/insurance/active'),
  policies: (limit = 50) => api.get<PolicyOut[]>(`/insurance/policies?limit=${limit}`),
  claims: (limit = 50) => api.get<ClaimOut[]>(`/insurance/claims?limit=${limit}`),
};

File: backend/packages/common/src/schemas/insurance.py
"""Trade Insurance API Pydantic schemas."""
from datetime import datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class InsuranceQuoteRequest(BaseModel):
    account_id: UUID
    symbol: str
    side: str = Field(..., pattern="^(buy|sell)$")
    lots: Decimal = Field(gt=0, le=100)
    leverage: int = Field(default=100, ge=1, le=2000)
    stop_loss: Optional[Decimal] = None
    take_profit: Optional[Decimal] = None


class InsuranceTierQuote(BaseModel):
    tier: str
    fee: float
    coverage_pct: float
    max_cap: float
    estimated_refund: float
    risk_score: float


class InsuranceActivateRequest(BaseModel):
    # Tier is the admin-defined label string (e.g. "50%", "70%"). The
    # legacy 4-tier regex (basic/advanced/pro/elite) was dropped during
    # the 2026-05-25 cleanup — any non-empty label up to 32 chars is OK.
    position_id: UUID
    tier: str = Field(..., min_length=1, max_length=32)


class InsuranceActivateResponse(BaseModel):
    policy_id: UUID
    fee_charged: Decimal
    status: str


class InsurancePolicyOut(BaseModel):
    id: UUID
    position_id: Optional[UUID] = None
    instrument_symbol: Optional[str] = None
    tier: str
    fee: Decimal
    coverage_pct: Decimal
    max_cap: Decimal
    status: str
    activated_at: datetime
    settled_at: Optional[datetime] = None


class InsuranceClaimOut(BaseModel):
    id: UUID
    policy_id: UUID
    loss_amount: Decimal
    claim_amount: Decimal
    paid_at: datetime
